src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -7,400 +7,457 @@
+  /*@volatile*/ private[this] var _result: Try[T] = null
+  final def result: Try[T] = _result
+  override protected def tryAcquireShared(ignored: Int): Int = if (getState != 0) 1 else -1
+  override protected def tryReleaseShared(ignore: Int): Boolean = {
+    setState(1)
+    true
+  }
+  override def apply(value: Try[T]): Unit = {
+    _result = value // This line MUST go before releaseShared
+    releaseShared(1)
+  }
+}
 
-private[concurrent] trait Promise[T] extends scala.concurrent.Promise[T] with scala.concurrent.Future[T] {
-  def future: this.type = this
-
-  import scala.concurrent.Future
-  import scala.concurrent.impl.Promise.DefaultPromise
+private[concurrent] final object Promise {
+    /**
+     * Link represents a completion dependency between 2 DefaultPromises.
+     * As the DefaultPromise referred to by a Link can itself be linked to another promise
+     * `relink` traverses such chains and compresses them so that the link always points
+     * to the root of the dependency chain.
+     *
+     * In order to conserve memory, the owner of a Link (a DefaultrPomise) is not stored
+     * on the Link, but is instead passed in as a parameter to the operation(s).
+     *
+     * If when compressing a chain of Links it is discovered that the root has been completed,
+     * the `owner`'s value is completed with that value, and the Link chain is discarded.
+     **/
+    private final class Link[T](to: DefaultPromise[T]) extends AtomicReference[DefaultPromise[T]](to) {
+      /**
+       * Compresses this chain and returns the currently known root of this chain of Links.
+       **/
+      final def promise(owner: DefaultPromise[T]): DefaultPromise[T] =
+        relink(target = this.get(), owner = owner)
+
+      /**
+       * Traverses the Link chain to find the current root.
+       * Unlinks (removes) the link chain if the root is discovered to be already completed,
+       * and completes the `owner` with that result.
+       **/
+      @tailrec private[this] final def root(of: DefaultPromise[T], owner: DefaultPromise[T]): DefaultPromise[T] = {
+        val value = of.get()
+        if (value.isInstanceOf[Link[T]]) root(of = value.asInstanceOf[Link[T]].get(), owner = owner)
+        else if ((owner ne null) && value.isInstanceOf[Try[T]]) { // Link completion detected, unlink.
+          unlink(owner = owner, value = value.asInstanceOf[Try[T]])
+          owner
+        } else of
+      }
 
-  override def transform[S](f: Try[T] => Try[S])(implicit executor: ExecutionContext): Future[S] = {
-    val p = new DefaultPromise[S]()
-    onComplete { result => p.complete(try f(result) catch { case NonFatal(t) => Failure(t) }) }
-    p.future
-  }
+      /**
+       * The combination of traversing and possibly unlinking of a given `target` DefaultPromise.
+       **/
+      @tailrec final def relink(target: DefaultPromise[T], owner: DefaultPromise[T]): DefaultPromise[T] = {
+        val current = get()
```

Comment:
@richdougherty I know. It was either that or introducing yet-another-branch :/

Reply:
OK, fair enough. :) Where would the branch be? Do you want to avoid the branch because it affects performance or because it makes the code more complicated?

Reply:
@richdougherty Both. :)

Reply:
@richdougherty I've adressed this now :)